File: src/types/boardReaction.ts
import {ReactionType} from "./reaction";

/**
 * representation of a board reaction on the server side
 */
export type BoardReactionType = {
  id: string;
  user: string;
  reactionType: ReactionType;
};

export const BOARD_REACTION_EMOJI_MAP = new Map<ReactionType, string>([
  ["tada", "🎉"],
  ["applause", "👏"],
  ["heart", "💖"],
  ["like", "👍"],
  ["dislike", "👎"],
]);

export type BoardReactionState = BoardReactionType[];

File: server/src/database/types/board_reaction.go
package types

import (
	"encoding/json"
	"errors"
)

type BoardReaction string

const (
	BoardReactionTada     BoardReaction = "tada"
	BoardReactionApplause BoardReaction = "applause"
	BoardReactionHeart    BoardReaction = "heart"
	BoardReactionLike     BoardReaction = "like"
	BoardReactionDislike  BoardReaction = "dislike"
	BoardReactionStan     BoardReaction = "stan"
)

func (reaction *BoardReaction) UnmarshalJSON(b []byte) error {
	var s string
	err := json.Unmarshal(b, &s)
	if err != nil {
		return err
	}
	unmarshalledReaction := BoardReaction(s)
	switch unmarshalledReaction {
	case BoardReactionTada, BoardReactionApplause, BoardReactionHeart, BoardReactionLike, BoardReactionDislike, BoardReactionStan:
		*reaction = unmarshalledReaction
		return nil
	}
	return errors.New("invalid board reaction")
}
